### PR TITLE
Issue/821/access variable by xpath json

### PR DESCRIPTION
**Pull request by tschlat**
Added the feature requested with issue #821. 

I have implemented the feature into the changes of #820. 

The code got more complex than I thought it would get. The reason is the separation of JsonPath/Xpath functionality into different modules. So I wasn't able to access the JSonPath/Xpath utillities from the api-module that hosts the TestContext. I had to decouple the extraction mechanism for JsonPath and Xpath into the modules. Although there is a bunch of new files, the implemented approach pretty much follows the decoupling mechanisms already used thought the framework.

There is one toDo in com.consol.citrus.xml.XpathSegmentVariableExtractor  for which I request advice from you @christophd.

### CODE REVIEW

File: core/citrus-api/src/main/java/com/consol/citrus/variable/SegmentVariableExtractorRegistry.java
Comment (bbortt) [inline core/citrus-api/src/main/java/com/consol/citrus/variable/SegmentVariableExtractorRegistry.java:27]: use `List.of` for instantiation?

File: core/citrus-api/src/main/java/com/consol/citrus/variable/SegmentVariableExtractor.java
Comment (bbortt) [inline core/citrus-api/src/main/java/com/consol/citrus/variable/SegmentVariableExtractor.java:38]: write in lowercase letters, or is it commonly written in uppercase in this module? (there are some other `Logger` instaces as well in this changeset..)

File: core/citrus-api/src/main/java/com/consol/citrus/variable/VariableExpressionSegmentMatcher.java
Comment (bbortt) [inline core/citrus-api/src/main/java/com/consol/citrus/variable/VariableExpressionSegmentMatcher.java:47]: this is the group index of the regex pattern match?

File: validation/citrus-validation-json/src/test/java/com/consol/citrus/integration/JsonPathSegmentVariableExtractorIT.java
Comment (bbortt) [inline validation/citrus-validation-json/src/test/java/com/consol/citrus/integration/JsonPathSegmentVariableExtractorIT.java:10]: I just had an idea.. could we have the `@Ignore` of the "no empty test methods" sonar rule on the `@CitrusXmlTest` annotation? :D

File: validation/citrus-validation-xml/src/main/java/com/consol/citrus/xml/XpathSegmentVariableExtractor.java
Comment (bbortt) [inline validation/citrus-validation-xml/src/main/java/com/consol/citrus/xml/XpathSegmentVariableExtractor.java:24]: that is pretty obvious 😉

File: validation/citrus-validation-xml/src/main/java/com/consol/citrus/xml/XpathSegmentVariableExtractor.java
Comment (bbortt) [inline validation/citrus-validation-xml/src/main/java/com/consol/citrus/xml/XpathSegmentVariableExtractor.java:57]: did you think about matching only elements in namespaces? that'd be pretty clean.. but probably the "rolls-royce solution".. having this in the query string too might make it too unreadable.

File: validation/citrus-validation-xml/src/main/java/com/consol/citrus/xml/XpathSegmentVariableExtractor.java
Comment (bbortt) [inline validation/citrus-validation-xml/src/main/java/com/consol/citrus/xml/XpathSegmentVariableExtractor.java:64]: 🐔

**Review by bbortt — CHANGES_REQUESTED** [review]

**Comment by bbortt** [comment]
@tschlat pretty cool enhancement! but [the build](https://github.com/citrusframework/citrus/runs/4549698732?check_suite_focus=true) didn't like your changes..

**Review by tschlat — COMMENTED** [review]
Included feedback of bbortt.

File: core/citrus-api/src/main/java/com/consol/citrus/variable/SegmentVariableExtractor.java
Comment (christophd) [reply core/citrus-api/src/main/java/com/consol/citrus/variable/SegmentVariableExtractor.java]: yes loggers usage needs to be unified in the whole framework. I think they all should be `private static final`. And by convention static final members are spelled in uppercase. Would be happy to have this addressed in a separate PR

**Review by christophd — COMMENTED** [review]
This is really a great enhancement! Many thanks! I like the configuration via resource lookup, well done!

I'd just argue the document caching mechanism and we need to also add this to Spring based configuration

File: core/citrus-api/src/main/java/com/consol/citrus/variable/SegmentVariableExtractor.java
Comment (tschlat) [reply core/citrus-api/src/main/java/com/consol/citrus/variable/SegmentVariableExtractor.java]: I have added an issue. We can discuss the final definition for the log there.
https://github.com/citrusframework/citrus/issues/829

**Comment by tschlat** [comment]
I have added a commit with the needed changes identified by @christophd

**Review by tschlat — COMMENTED** [review]
I have added fixes for your review @christophd

**Review by christophd — APPROVED** [review]

**Comment by christophd** [comment]
LGTM Many thanks! Unit tests in CI job failing though

**Comment by tschlat** [comment]
@christophd please check changes of this [commit](https://github.com/citrusframework/citrus/pull/827/commits/52540a9ab9c76317209b2ab34f25cb8df0023744) which I provided as fix to [issue#831](https://github.com/citrusframework/citrus/issues/831)

**Comment by tschlat** [comment]
@christophd please carefully check changes on this [commit](https://github.com/citrusframework/citrus/pull/827/commits/6896b956084fced3af34e1ffcfa81716bfb159c5) which I provided as fix to [issue#832](https://github.com/citrusframework/citrus/issues/832).

File: core/citrus-base/src/test/java/com/consol/citrus/functions/core/ReadFileResourceFunctionTest.java
Comment (bbortt) [inline core/citrus-base/src/test/java/com/consol/citrus/functions/core/ReadFileResourceFunctionTest.java:44]: I thought `List.of` is better use.. but I cannot find sources :(

File: core/citrus-api/src/main/java/com/consol/citrus/util/IsJsonPredicate.java
Comment (bbortt) [inline core/citrus-api/src/main/java/com/consol/citrus/util/IsJsonPredicate.java:11]: add a private constructor, this will prevent anyone from unintentionally creating a new instance. I personally prefer accessors too (aka `static getInstance()`), but that's probably a matter of taste.

File: core/citrus-api/src/main/java/com/consol/citrus/util/IsXmlPredicate.java
Comment (bbortt) [inline core/citrus-api/src/main/java/com/consol/citrus/util/IsXmlPredicate.java:10]: add a private constructor, this will prevent anyone from unintentionally creating a new instance.

File: core/citrus-api/src/main/java/com/consol/citrus/variable/VariableExpressionIterator.java
Comment (bbortt) [inline core/citrus-api/src/main/java/com/consol/citrus/variable/VariableExpressionIterator.java:133]: that's a useless `super` invocation. the file hasn't been formatted too 😉

File: core/citrus-api/src/main/java/com/consol/citrus/variable/VariableExpressionSegmentMatcher.java
Comment (bbortt) [inline core/citrus-api/src/main/java/com/consol/citrus/variable/VariableExpressionSegmentMatcher.java:128]: I think you can ask for `matcher.results().count()` or something.

File: vintage/citrus-java-dsl/src/main/java/com/consol/citrus/dsl/builder/HttpActionBuilder.java
Comment (bbortt) [inline vintage/citrus-java-dsl/src/main/java/com/consol/citrus/dsl/builder/HttpActionBuilder.java:30]: thanks @tschlat! that's the right one - follow up for #817. my bad.

**Review by bbortt — CHANGES_REQUESTED** [review]
a review for christmas 🎄

File: core/citrus-base/src/test/java/com/consol/citrus/functions/core/ReadFileResourceFunctionTest.java
Comment (tschlat) [reply core/citrus-base/src/test/java/com/consol/citrus/functions/core/ReadFileResourceFunctionTest.java:44]: It boils down to: "it's a matter of choice."
https://dzone.com/articles/singleton-list-showdown-collectionssingletonlist-v

I am used to use singletonList because of java 1.8 - thats all. Since you brought up the matter i lookup into the code. I changed that piece of code because of a warning on the usage of Arrays.asList with only a single parameter. I can use List.of in both places of the code without a warning. This makes it consistent.

File: core/citrus-api/src/main/java/com/consol/citrus/util/IsJsonPredicate.java
Comment (tschlat) [reply core/citrus-api/src/main/java/com/consol/citrus/util/IsJsonPredicate.java:11]: private constructor is certainly required to fullfill singleton pattern. I also added a getInstance for your taste.

File: core/citrus-api/src/main/java/com/consol/citrus/util/IsXmlPredicate.java
Comment (tschlat) [reply core/citrus-api/src/main/java/com/consol/citrus/util/IsXmlPredicate.java:10]: See above.

File: core/citrus-api/src/main/java/com/consol/citrus/variable/VariableExpressionIterator.java
Comment (tschlat) [reply core/citrus-api/src/main/java/com/consol/citrus/variable/VariableExpressionIterator.java:133]: Removed that one. 

I also reformatted the code.

I now formatted the code using the standard settings of IntelliJ. Hope that meets your

File: core/citrus-api/src/main/java/com/consol/citrus/variable/VariableExpressionSegmentMatcher.java
Comment (tschlat) [reply core/citrus-api/src/main/java/com/consol/citrus/variable/VariableExpressionSegmentMatcher.java:128]: That's nice. I was actually looking for something like that. Obviously not eagerly enough.

**Comment by tschlat** [comment]
@bbortt I accepted all your suggestions. 

@christophd if you could have a look at the two commits concerning other issues and could kick the workflow we could eventually make it for christmas.

File: core/citrus-api/src/main/java/com/consol/citrus/util/IsJsonPredicate.java
Comment (bbortt) [reply core/citrus-api/src/main/java/com/consol/citrus/util/IsJsonPredicate.java]: clap

**Review by bbortt — APPROVED** [review]
/lgtm, many thanks.

File: validation/citrus-validation-xml/src/main/java/com/consol/citrus/xml/XpathSegmentVariableExtractor.java
Comment (bbortt) [reply validation/citrus-validation-xml/src/main/java/com/consol/citrus/xml/XpathSegmentVariableExtractor.java]: I meant it as "nit picking" 😉 because the line was not formatted properly..

**Comment by christophd** [comment]
@tschlat sorry, did not make it before the holidays. just started the workflow

File: core/citrus-base/src/main/java/com/consol/citrus/functions/core/ReadFileResourceFunction.java
Comment (christophd) [inline core/citrus-base/src/main/java/com/consol/citrus/functions/core/ReadFileResourceFunction.java:53]: I think replacing the dynamic content by default is problematic here. This is because the file content can also be non textual e.g. an image. The dynamic replace of variables in this case will corrupt the content or fail miserably. 

This is why the base64 part of the function has skipped the replacement of variables because reading files in base64 has been used with images almost all the time.

What about introducing another optional function parameter to explicitly enable variable replacement? This could parameter could default to "true" so users can skip the replacement on demand.

File: core/citrus-base/src/main/java/com/consol/citrus/functions/core/ReadFileResourceFunction.java
Comment (christophd) [inline core/citrus-base/src/main/java/com/consol/citrus/functions/core/ReadFileResourceFunction.java:69]: How about using a method argument `filePath` instead of the whole parameter list? Only the first entry in the list (the file path) is used in the method so no reason to pass in the whole list as an argument.

**Review by christophd — COMMENTED** [review]
I see a problem with non textual file content and the variable replacement. I have described in more detail in the comments

File: vintage/citrus-java-dsl/src/main/java/com/consol/citrus/dsl/builder/HttpActionBuilder.java
Comment (christophd) [reply vintage/citrus-java-dsl/src/main/java/com/consol/citrus/dsl/builder/HttpActionBuilder.java:30]: yes, that's the right place. 👍 

@bbortt what about reverting the changes made in #817? I think the builder should only accept HttpServer endpoints. I need to review the use case in the citrus-simulator to understand why another endpoint type is required here.

It is ok to provide the Endpoint option in the vintage modules but I think using the builder with generic endpoint type is not the way we should go in the long run.

File: core/citrus-base/src/main/java/com/consol/citrus/functions/core/ReadFileResourceFunction.java
Comment (tschlat) [reply core/citrus-base/src/main/java/com/consol/citrus/functions/core/ReadFileResourceFunction.java:53]: I put the default for the new parameter to false to ensure backwards compatibility.
As you noted: My test is probably the only piece of code that makes use of this feature.

**Comment by tschlat** [comment]
@christophd magst Du noch mal meinen Change für #831 reviewen und den Workflow noch mal triggern?

**Comment by bbortt** [comment]
@christophd maybe one of us (@tschlat or me) can be maintainer here as well as in citrusframework/citrus-simulator? would be faster for us, if we could trigger the workflow ourselfs.

**Comment by christophd** [comment]
@bbortt @tschlat we have a green build! 🥳  @tschlat do you still want to squash and reorganize the commits before merge?

File: endpoints/citrus-ws/src/main/java/com/consol/citrus/ws/message/SoapAttachment.java
Comment (christophd) [inline endpoints/citrus-ws/src/main/java/com/consol/citrus/ws/message/SoapAttachment.java:260]: is this a leftover from debugging session?

**Comment by tschlat** [comment]
> @bbortt @tschlat we have a green build! 🥳 @tschlat do you still want to squash and reorganize the commits before merge?

This has already been done. I have a commit per issue. Any objections?

**Comment by tschlat** [comment]
@christophd could you make me a maintainer so that I can trigger the build myself?

**Comment by bbortt** [comment]
> @christophd could you make me a maintainer so that I can trigger the build myself?

I can help you.. just ping me 😉

**Comment by bbortt** [comment]
thanks @christophd 🚀

**Comment by tschlat** [comment]
Yes thanks a lot for the support @christophd

**Comment by tschlat** [comment]
and @bbortt

File: vintage/citrus-java-dsl/src/main/java/com/consol/citrus/dsl/builder/HttpActionBuilder.java
Comment (bbortt) [reply vintage/citrus-java-dsl/src/main/java/com/consol/citrus/dsl/builder/HttpActionBuilder.java:30]: see https://github.com/citrusframework/citrus/pull/840.

**Pull request merged**
